File: FarNet/FarNetMan/Viewer.cpp
#include "StdAfx.h"
#include "Viewer.h"
#include "Viewer0.h"
#include "Window.h"

namespace FarNet
{
void AnyViewer::ViewText(String^ text, String^ title, OpenMode mode)
{
	Works::EditorTools::ViewText(text, title, mode);
}

Viewer::Viewer()
: _id(-1)
, _Title(String::Empty)
, _CodePage(CP_DEFAULT)
, _Window(0, 0, -1, -1)
{}

void Viewer::Open(OpenMode mode)
{
	AssertClosed();

	// flags
	int flags = 0;

	if (_DisableHistory)
		flags |= VF_DISABLEHISTORY;

	switch(_Switching)
	{
	case FarNet::Switching::Enabled:
		flags |= VF_ENABLE_F6;
		break;
	case FarNet::Switching::Auto:
		if (_DeleteSource == FarNet::DeleteSource::None &&
			_Closed == nullptr &&
			_GotFocus == nullptr &&
			_LosingFocus == nullptr &&
			_Opened == nullptr)
			flags |= VF_ENABLE_F6;
		break;
	}

	switch(mode)
	{
	case OpenMode::None:
		flags |= (VF_NONMODAL | VF_IMMEDIATERETURN); break;
	case OpenMode::Wait:
		flags |= VF_NONMODAL; break;
	}

	switch(_DeleteSource)
	{
	case FarNet::DeleteSource::UnusedFile:
		flags |= VF_DELETEONLYFILEONCLOSE; break;
	case FarNet::DeleteSource::UnusedFolder:
		flags |= VF_DELETEONCLOSE; break;
	}

	PIN_ES(pinFileName, _FileName);
	PIN_ES(pinTitle, _Title);

	// from modal? set modal
	bool preIsModal = FarNet::Window::Instance.IsModal;
	if (preIsModal)
		flags &= ~VF_NONMODAL;

	// open: see editor
	_id = -1;
	Viewer0::_viewerWaiting = this;
	Info.Viewer(
		pinFileName,
		pinTitle,
		_Window.Left,
		_Window.Top,
		_Window.Right,
		_Window.Bottom,
		flags,
		_CodePage); //?? test window values

	// errors: see editor
	if (_id == -1)
		throw gcnew InvalidOperationException("Cannot open the file '" + FileName + "'");
}

IntPtr Viewer::Id::get()
{
	return (IntPtr)_id;
}

DeleteSource Viewer::DeleteSource::get()
{
	return _DeleteSource;
}

void Viewer::DeleteSource::set(FarNet::DeleteSource value)
{
	_DeleteSource = value;
}

Switching Viewer::Switching::get()
{
	return _Switching;
}

void Viewer::Switching::set(FarNet::Switching value)
{
	AssertClosed();
	_Switching = value;
}

bool Viewer::DisableHistory::get()
{
	return _DisableHistory;
}

void Viewer::DisableHistory::set(bool value)
{
	AssertClosed();
	_DisableHistory = value;
}

bool Viewer::IsOpened::get()
{
	return _id >= 0;
}

String^ Viewer::FileName::get()
{
	return _FileName;
}

void Viewer::FileName::set(String^ value)
{
	AssertClosed();
	_FileName = value;
}

//! Viewer does not have VCTL_GETTITLE
String^ Viewer::Title::get()
{
	return String::IsNullOrEmpty(_Title) ? _FileName : _Title;
}

//! Viewer does not have VCTL_SETTITLE
void Viewer::Title::set(String^ value)
{
	AssertClosed();
	_Title = value;
}

Int64 Viewer::FileSize::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	if (vi.ViewerID >= 0 && vi.ViewerID == _id)
		return vi.FileSize;
	else
		return -1;
}

Point Viewer::WindowSize::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	Point r;
	if (vi.ViewerID >= 0 && vi.ViewerID == _id)
	{
		r.X = (int)vi.WindowSizeX;
		r.Y = (int)vi.WindowSizeY;
	}
	return r;
}

Place Viewer::Window::get()
{
	return _Window;
}

void Viewer::Window::set(Place value)
{
	AssertClosed();
	_Window = value;
}

ViewFrame Viewer::Frame::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	ViewFrame r;
	if (vi.ViewerID >= 0 && vi.ViewerID == _id)
	{
		r.Offset = vi.FilePos;
		r.Column = vi.LeftPos;
	}
	return r;
}

void Viewer::Frame::set(ViewFrame value)
{
	AssertCurrentViewer();
	ViewerSetPosition vsp = {sizeof(vsp)};
	vsp.Flags = VSP_NORETNEWPOS;
	vsp.LeftPos = value.Column;
	vsp.StartPos = value.Offset;
	Info.ViewerControl(_id, VCTL_SETPOSITION, 0, &vsp);
}

Int64 Viewer::SetFrame(Int64 pos, int left, ViewFrameOptions options)
{
	AssertCurrentViewer();
	ViewerSetPosition vsp = {sizeof(vsp)};
	vsp.Flags = (DWORD)options;
	vsp.LeftPos = left;
	vsp.StartPos = pos;
	Info.ViewerControl(_id, VCTL_SETPOSITION, 0, &vsp);
	return vsp.StartPos;
}

void Viewer::Close()
{
	AssertCurrentViewer();
	Info.ViewerControl(_id, VCTL_QUIT, 0, 0);
}

void Viewer::Redraw()
{
	AssertCurrentViewer();
	Info.ViewerControl(_id, VCTL_REDRAW, 0, 0);
}

void Viewer::SelectText(Int64 symbolStart, int symbolCount)
{
	AssertCurrentViewer();
	if (symbolCount <= 0)
	{
		Info.ViewerControl(_id, VCTL_SELECT, 0, 0);
	}
	else
	{
		ViewerSelect vs = {sizeof(vs)};
		vs.BlockLen = symbolCount;
		vs.BlockStartPos = symbolStart;
		Info.ViewerControl(_id, VCTL_SELECT, 0, &vs);
	}
}

void Viewer::AssertClosed()
{
	if (IsOpened) throw gcnew InvalidOperationException("Viewer must not be open for this operation.");
}

ViewerViewMode Viewer::ViewMode::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	if (vi.ViewerID < 0 || vi.ViewerID != _id)
		return ViewerViewMode::Text;
	else
		return (ViewerViewMode)vi.CurMode.ViewMode;
}
void Viewer::ViewMode::set(ViewerViewMode value)
{
	AssertCurrentViewer();
	ViewerSetMode vsm = {sizeof(vsm)};
	vsm.Flags = 0;
	vsm.Type = VSMT_VIEWMODE;
	vsm.iParam = (intptr_t)value;
	Info.ViewerControl(_id, VCTL_SETMODE, 0, &vsm);
}

bool Viewer::WrapMode::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	if (vi.ViewerID < 0 || vi.ViewerID != _id)
		return false;
	else
		return (vi.CurMode.Flags & VMF_WRAP) != 0;
}
void Viewer::WrapMode::set(bool value)
{
	AssertCurrentViewer();
	ViewerSetMode vsm = {sizeof(vsm)};
	vsm.Flags = 0;
	vsm.Type = VSMT_WRAP;
	vsm.iParam = value;
	Info.ViewerControl(_id, VCTL_SETMODE, 0, &vsm);
}

bool Viewer::WordWrapMode::get()
{
	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	if (vi.ViewerID < 0 || vi.ViewerID != _id)
		return false;
	else
		return (vi.CurMode.Flags & VMF_WORDWRAP) != 0;
}
void Viewer::WordWrapMode::set(bool value)
{
	AssertCurrentViewer();
	ViewerSetMode vsm = {sizeof(vsm)};
	vsm.Flags = 0;
	vsm.Type = VSMT_WORDWRAP;
	vsm.iParam = value;
	Info.ViewerControl(_id, VCTL_SETMODE, 0, &vsm);
}

int Viewer::CodePage::get()
{
	if (!IsOpened)
		return (int)_CodePage;

	ViewerInfo vi; ViewerControl_VCTL_GETINFO(vi, true);
	if (vi.ViewerID < 0 || vi.ViewerID != _id)
		return 0;

	return (int)vi.CurMode.CodePage;
}

void Viewer::CodePage::set(int value)
{
	AssertClosed();

	_CodePage = value;
}

DateTime Viewer::TimeOfGotFocus::get()
{
	return _TimeOfGotFocus;
}

DateTime Viewer::TimeOfOpen::get()
{
	return _TimeOfOpen;
}

}

File: FarNet/FarNet/Viewer.cs

/*
FarNet plugin for Far Manager
Copyright (c) 2006-2015 Roman Kuzmin
*/

using System;
using System.Diagnostics;

namespace FarNet
{
	/// <summary>
	/// Common viewer events.
	/// </summary>
	public abstract class IViewerBase
	{
		/// <summary>
		/// Called when the viewer is closed.
		/// </summary>
		/// <remarks>
		/// This event is called once for the viewer instance, even if there were several files opened in it,
		/// e.g. on [Add], [Subtract] keys the <see cref="Opened"/> is called every time.
		/// <para>
		/// Don't operate on the viewer, it has really gone.
		/// </para>
		/// </remarks>
		public abstract event EventHandler Closed; // [_100117_101226]
		/// <summary>
		/// Called when a file is opened in the viewer.
		/// </summary>
		/// <remarks>
		/// This event can be called more than once for the same viewer instance,
		/// e.g. on [Add], [Subtract] keys.
		/// </remarks>
		public abstract event EventHandler Opened;
		/// <summary>
		/// Called when the viewer window has got focus.
		/// </summary>
		public abstract event EventHandler GotFocus;
		/// <summary>
		/// Called when the viewer window is losing focus.
		/// </summary>
		public abstract event EventHandler LosingFocus;
	}

	/// <summary>
	/// Any viewer operator.
	/// Exposed as <see cref="IFar.AnyViewer"/>.
	/// </summary>
	public abstract class IAnyViewer : IViewerBase
	{
		/// <summary>
		/// Opens a viewer to view some text.
		/// </summary>
		/// <param name="text">The text to view.</param>
		/// <param name="title">The viewer title.</param>
		/// <param name="mode">The open mode.</param>
		public abstract void ViewText(string text, string title, OpenMode mode);
	}

	/// <summary>
	/// Viewer operator. Exposed as <see cref="IFar.Viewer"/>. Created by <see cref="IFar.CreateViewer"/>.
	/// </summary>
	/// <remarks>
	/// Normally this object should be created or requested, used instantly and never kept for future use.
	/// When you need the current viewer operator next time call <see cref="IFar.Viewer"/> again to get it.
	/// </remarks>
	public abstract class IViewer : IViewerBase
	{
		/// <summary>
		/// Gets the internal identifier.
		/// </summary>
		public abstract IntPtr Id { get; }
		/// <summary>
		/// Gets or sets the option to delete the source file on exit.
		/// </summary>
		public abstract DeleteSource DeleteSource { get; set; }
		/// <summary>
		/// Tells how editor\viewer switching should work on [F6].
		/// Set it before opening.
		/// </summary>
		public abstract Switching Switching { get; set; }
		/// <summary>
		/// Tells to not use history.
		/// Set it before opening.
		/// </summary>
		public abstract bool DisableHistory { get; set; }
		/// <summary>
		/// Name of a file being viewed. Set it before opening.
		/// On opening it can be corrected, e.g. converted into full path.
		/// </summary>
		public abstract string FileName { get; set; }
		/// <summary>
		/// Gets or sets the code page identifier.
		/// </summary>
		/// <remarks>
		/// Before opening it sets encoding for reading a file.
		/// After opening it only gets the current encoding.
		/// </remarks>
		public abstract int CodePage { get; set; }
		/// <summary>
		/// Gets or sets the start window place.
		/// Set it before opening.
		/// </summary>
		public abstract Place Window { get; set; }
		/// <summary>
		/// Gets the current window size.
		/// </summary>
		public abstract Point WindowSize { get; }
		/// <summary>
		/// Gets or sets the window title.
		/// Set it before opening.
		/// </summary>
		public abstract string Title { get; set; }
		/// <summary>
		/// Opens the viewer.
		/// </summary>
		/// <remarks>
		/// It is the same as <see cref="Open(OpenMode)"/> with open mode <see cref="OpenMode.None"/>.
		/// See remarks there.
		/// </remarks>
		public void Open()
		{
			Open(OpenMode.None);
		}
		/// <summary>
		/// Opens the viewer.
		/// </summary>
		/// <param name="mode">The open mode.</param>
		/// <remarks>
		/// To open a viewer you should create a viewer operator by <see cref="IFar.CreateViewer"/>,
		/// set at least its <see cref="FileName"/> and optionally: <see cref="DeleteSource"/>,
		/// <see cref="DisableHistory"/>, <see cref="Switching"/>, <see cref="Title"/>, and
		/// <see cref="Window"/>. Then this method is called.
		/// </remarks>
		public abstract void Open(OpenMode mode);
		/// <summary>
		/// Gets the current file size, in symbols, not in bytes.
		/// </summary>
		public abstract long FileSize { get; }
		/// <summary>
		/// Gets the current view frame.
		/// </summary>
		public abstract ViewFrame Frame { get; set; }
		/// <summary>
		/// Sets the current view frame.
		/// </summary>
		/// <param name="offset">New file position (depends on options).</param>
		/// <param name="column">New left position.</param>
		/// <param name="options">Options.</param>
		/// <returns>New actual position.</returns>
		public abstract long SetFrame(long offset, int column, ViewFrameOptions options);
		/// <summary>
		/// Closes the current viewer window.
		/// </summary>
		public abstract void Close();
		/// <summary>
		/// Redraws the current viewer window.
		/// </summary>
		public abstract void Redraw();
		/// <summary>
		/// Selects the block of text in the current viewer.
		/// </summary>
		/// <param name="symbolStart">Selection start in charactes, not in bytes.</param>
		/// <param name="symbolCount">Selected character count.</param>
		public abstract void SelectText(long symbolStart, int symbolCount);
		/// <summary>
		/// Gets or sets the view mode in the current viewer (~ [F4]).
		/// </summary>
		public abstract ViewerViewMode ViewMode { get; set; }
		/// <summary>
		/// Gets or sets the wrap mode in the current editor (~ [F2]).
		/// </summary>
		public abstract bool WrapMode { get; set; }
		/// <summary>
		/// Gets or sets the word wrap mode in the current editor (~ [ShiftF2]).
		/// </summary>
		public abstract bool WordWrapMode { get; set; }
		/// <summary>
		/// Gets the opening time of the instance.
		/// </summary>
		public abstract DateTime TimeOfOpen { get; }
		/// <summary>
		/// Makes the instance window active.
		/// </summary>
		/// <remarks>It may throw if the window cannot be activated.</remarks>
		public abstract void Activate();
	}

	/// <summary>
	/// Viewer modes.
	/// </summary>
	public enum ViewerViewMode
	{
		/// <summary>
		/// Text view mode.
		/// </summary>
		Text,
		/// <summary>
		/// Hex view mode.
		/// </summary>
		Hex,
		/// <summary>
		/// Dump view mode.
		/// </summary>
		Dump
	}

	/// <summary>
	/// Options for <see cref="IViewer.SetFrame"/>.
	/// </summary>
	[Flags]
	public enum ViewFrameOptions
	{
		///
		None,
		/// <summary>
		/// Don't redraw.
		/// </summary>
		NoRedraw = 1,
		/// <summary>
		/// Offset is defined in percents.
		/// </summary>
		Percent = 2,
		/// <summary>
		/// Offset is relative to the current (and can be negative).
		/// </summary>
		Relative = 4,
	}

	/// <summary>
	/// Open modes of editor and viewer.
	/// </summary>
	public enum OpenMode
	{
		/// <summary>
		/// Tells to open not modal editor or viewer and return immediately.
		/// </summary>
		None,
		/// <summary>
		/// Tells to open not modal editor or viewer and wait for exit.
		/// </summary>
		Wait,
		/// <summary>
		/// Tells to open modal editor or viewer.
		/// </summary>
		Modal,
	}

	/// <summary>
	/// Options to delete temporary source files and empty folders after use.
	/// </summary>
	public enum DeleteSource
	{
		/// <summary>
		/// Default action: do not delete a file.
		/// </summary>
		None,
		/// <summary>
		/// Try to delete a file always. It is not recommended if editor\viewer switching is enabled (F6).
		/// You may set it at any time, i.e. before or after opening.
		/// </summary>
		File,
		/// <summary>
		/// The same as <see cref="File"/> plus delete its folder if it is empty.
		/// You may set it at any time, i.e. before or after opening.
		/// </summary>
		Folder,
		/// <summary>
		/// Delete a file if it was not used. The file is used if:
		/// *) it was saved;
		/// *) there was editor\viewer switching (F6).
		/// *) it is opened in another editor or viewer.
		/// You should set it before opening.
		/// </summary>
		UnusedFile,
		/// <summary>
		/// The same as <see cref="UnusedFile"/> plus delete its folder if it is empty.
		/// You should set it before opening.
		/// </summary>
		UnusedFolder,
	}

	/// <summary>
	/// Viewer frame info.
	/// </summary>
	public struct ViewFrame
	{
		/// <param name="offset">See <see cref="Offset"/></param>
		/// <param name="column">See <see cref="Column"/></param>
		public ViewFrame(long offset, long column)
			: this()
		{
			Offset = offset;
			Column = column;
		}
		/// <summary>
		/// Offset in the file.
		/// </summary>
		public long Offset { get; set; }
		/// <summary>
		/// Leftmost visible column index.
		/// </summary>
		public long Column { get; set; }
		/// <include file='doc.xml' path='doc/OpEqual/*'/>
		public static bool operator ==(ViewFrame left, ViewFrame right)
		{
			return
				left.Offset == right.Offset &&
				left.Column == right.Column;
		}
		/// <include file='doc.xml' path='doc/OpNotEqual/*'/>
		public static bool operator !=(ViewFrame left, ViewFrame right)
		{
			return !(left == right);
		}
		/// <inheritdoc/>
		public override bool Equals(object obj)
		{
			return obj != null && obj.GetType() == typeof(ViewFrame) && this == (ViewFrame)obj;
		}
		/// <inheritdoc/>
		public override int GetHashCode()
		{
			return (int)Offset | ((int)Column << 16);
		}
		/// <summary>
		/// Returns the string "(Offset, Column)".
		/// </summary>
		public override string ToString()
		{
			return "(" + Offset + ", " + Column + ")";
		}
	}

}
